src/tools/ceph_dedup_tool.cc

```diff
@@ -1144,49 +1144,58 @@ int make_dedup_object(const std::map < std::string, std::string > &opts,
      * the object a manifest object, the tier_flush() will remove
      * it and replace it with the real contents.
      */
-    // convert object to manifest object
-    ObjectWriteOperation op;
-    bufferlist temp;
-    temp.append("temp");
-    op.write_full(temp);
-
-    auto gen_r_num = [] () -> string {
-      std::random_device rd;
-      std::mt19937 gen(rd());
-      std::uniform_int_distribution<uint64_t> dist;
-      uint64_t r_num = dist(gen);
-      return to_string(r_num);
-    };
-    string temp_oid = gen_r_num();
-    // create temp chunk object for set-chunk
-    ret = chunk_io_ctx.operate(temp_oid, &op);
-    if (ret == -EEXIST) {
-      // one more try
-      temp_oid = gen_r_num();
-      ret = chunk_io_ctx.operate(temp_oid, &op);
-    }
-    if (ret < 0) {
-      cerr << " operate fail : " << cpp_strerror(ret) << std::endl;
-      goto out;
-    }
-
-    // set-chunk to make manifest object
-    ObjectReadOperation chunk_op;
-    chunk_op.set_chunk(0, 4, chunk_io_ctx, temp_oid, 0,
-      CEPH_OSD_OP_FLAG_WITH_REFERENCE);
-    ret = io_ctx.operate(object_name, &chunk_op, NULL);
-    if (ret < 0) {
-      cerr << " set_chunk fail : " << cpp_strerror(ret) << std::endl;
-      goto out;
-    }
-
     // tier-flush to perform deduplication
     ObjectReadOperation flush_op;
     flush_op.tier_flush();
     ret = io_ctx.operate(object_name, &flush_op, NULL);
-    if (ret < 0) {
+    if (ret < 0 && ret != -EPERM) {
       cerr << " tier_flush fail : " << cpp_strerror(ret) << std::endl;
       goto out;
+    } else if (ret == -EPERM) {
+      // this object is not manifest 
+      // try to convert object to manifest object
+      ObjectWriteOperation op;
+      bufferlist temp;
+      temp.append("temp");
```

Comment:
This still doesn't make sense to me.  Why do we need to write a temporary object in order to convert an object to a manifest object?

Reply:
This is...very unintuitive.  Is there a reason why set_chunk() can't implicitly make an object a manifest object?

Reply:
This code seems to:
1) Create a temporary object with a random name with contents "temp"
2) calls set_chunk to set the first 4 bytes of the target object to point to this temp object -- note that this won't generally match the first 4 bytes of the existing object
3) calls tier_flush to...re-dedup the object contents back to the base pool

Is that right?  I suppose a read between 2 and 3 only doesn't return the wrong data ("temp") because the chunk_info_t for that chunk will be !is_missing()?

The "false-positive based reference counting design" doesn't seem to require any of the above.  It seems to be simply because tier_flush won't implicitly make an object a manifest object.  Right?  It seems like it would be simpler for TIER_FLUSH to implicitly make an object a manifest object rather than simply returning 0.

Reply:
On second thought, it would be much simpler if we allow TIER_FLUSH to implicitly make a manifest object (I thought that we need to classify manifest and non-manifest). Can you take a look updated commits?